### PR TITLE
Remote key escrow 

### PR DESCRIPTION
**Pull request by endophage**
Remote key escrow via a GRPC implementation of the trustmanager.Storage interface.

Still finishing it up but intend for Notary Signer to be able to act as an escrow backend with configurable APIs (so it can host either or both of: current key storage for server and/or the new GRPC trustmanager.Storage server).

By implementing the low level trustmanager.Storage interface, passwords, and therefore key encryption and decryption, still happen at the client, so compromise of the remote key storage does not equate to compromise of the escrowed keys.

Signed-off-by: David Lawrence <david.lawrence@docker.com> (github: endophage)

### CODE REVIEW

**Comment by endophage** [comment]
This is about as tested as it's going to get. The compiled .pb.go is hurting coverage but I'm not going to write tests for generated code.

File: trustmanager/errors.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+// ErrKeyNotFound is returned when the keystore fails to retrieve a specific key.
+type ErrKeyNotFound struct {
+	KeyID string
+}
+
+// ErrKeyNotFound is returned when the keystore fails to retrieve a specific key.
+func (err ErrKeyNotFound) Error() string {
+	return fmt.Sprintf("signing key not found: %s", err.KeyID)
+}
```
Comment (riyazdf) [inline trustmanager/errors.go]: thanks for consolidating this 👍

File: trustmanager/interfaces.go [hunk 2]
```diff
@@ -34,32 +32,11 @@ type Storage interface {
...
-
-// ErrKeyNotFound is returned when the keystore fails to retrieve a specific key.
-type ErrKeyNotFound struct {
-	KeyID string
-}
-
-// ErrKeyNotFound is returned when the keystore fails to retrieve a specific key.
-func (err ErrKeyNotFound) Error() string {
-	return fmt.Sprintf("signing key not found: %s", err.KeyID)
+// KeyInfo stores the role, path, and gun for a corresponding private key ID
```
Comment (riyazdf) [inline trustmanager/interfaces.go]: nit since you're moving things around: remove `path` from description

File: trustmanager/remoteks/client.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+// RemoteStore is a wrapper around the GRPC storage client, translating between
+// the Go and GRPC APIs.
+type RemoteStore struct {
+	client   StoreClient
+	location string
+}
+
+var _ trustmanager.Storage = &RemoteStore{}
```
Comment (riyazdf) [inline trustmanager/remoteks/client.go]: why is this necessary?  Is this due to the nature of GRPC?

File: trustmanager/remoteks/client.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+func (s *RemoteStore) Remove(fileName string) error {
+	fm := &FileNameMsg{
+		FileName: fileName,
+	}
+	_, err := s.client.Remove(context.Background(), fm)
+	return err
+}
+
+// Get returns the file content found at fileName relative to the base directory
+// of the file store. The path is cleaned before being made absolute to ensure
```
Comment (riyazdf) [inline trustmanager/remoteks/client.go]: GH is unfortunately collapsing many diffs and so I'm having trouble finding the cleaning logic - could you please point me to it?

File: trustmanager/remoteks/client.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	return bm.Data, nil
+}
+
+// ListFiles returns a list of paths relative to the base directory of the
+// filestore. Any of these paths must be retrievable via the
+// Storage.Get method.
+func (s *RemoteStore) ListFiles() []string {
+	logrus.Infof("listing files from %s", s.location)
+	fl, err := s.client.ListFiles(context.Background(), &google_protobuf.Empty{})
+	if err != nil {
```
Comment (riyazdf) [inline trustmanager/remoteks/client.go]: can we add a test case for empty `ListFiles()`?

File: trustmanager/remoteks/client.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		logrus.Errorf("error listing files from %s: %s", s.location, err.Error())
+		return nil
+	}
+	return fl.FileNames
+
+}
+
+// Location returns a human readable indication of where the storage is located.
+func (s *RemoteStore) Location() string {
```
Comment (riyazdf) [inline trustmanager/remoteks/client.go]: nit: can we add a test case for `Location` in case of any regressions?

File: cmd/escrow/config.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	"fmt"
+	"github.com/docker/notary"
+	"github.com/docker/notary/storage"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/trustmanager/remoteks"
+	"github.com/docker/notary/utils"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"net"
```
Comment (endophage) [inline cmd/escrow/config.go]: This is how gogland is arranging them for me :-/ I need to try breaking them up into our normal format and see what it does with new imports that get auto-added. If we're all having to constantly tweak the import list though, maybe we stop trying so hard to organize it. It looks nicer the way we've been doing it, but I think everyone also loves just having imports auto-managed.

File: cmd/escrow/config.toml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+[storage]
```
Comment (endophage) [inline cmd/escrow/config.toml]: I'd actually prefer converting everything to toml, I just didn't want to do it in this PR. I'll file an issue for it. Viper will make it little more than just rewriting the json files and updating our Dockerfiles and docker-compose files to point at the toml versions.

File: trustmanager/remoteks/client.go (hunk 3 above)
Comment (endophage) [inline trustmanager/remoteks/client.go]: Hacky go compile time check to ensure a type implements the desired interface. You'll see it around distribution a lot too. It's relatively idiomatic if unofficial.

File: trustmanager/remoteks/client.go (hunk 4 above)
Comment (endophage) [inline trustmanager/remoteks/client.go]: These comments are just copied from the interface definition. The cleaning actually happens in the server side after passing through to the real underlying storage.

File: trustmanager/remoteks/client.go (hunk 5 above)
Comment (endophage) [inline trustmanager/remoteks/client.go]: Already do :-) https://github.com/docker/notary/pull/1096/files/ca60db8420cc08ec94536811fcdcc48e789b7f96#diff-fc36b6ceeef2f9d048a363a6eeab3b0cR69

Comment (endophage) [inline trustmanager/remoteks/client.go]: The GRPC API for Go always adds an error, but ListFiles never actually errors. The only way this error case would be hit is an error at the GRPC level.

File: cmd/escrow/config.toml (hunk 8 above)
Comment (riyazdf) [inline cmd/escrow/config.toml]: sounds good to me

Comment (endophage) [inline cmd/escrow/config.toml]: https://github.com/docker/notary/issues/1102

File: Makefile [hunk 9]
```diff
@@ -176,6 +180,9 @@ client: ${PREFIX}/bin/notary
 binaries: ${PREFIX}/bin/notary-server ${PREFIX}/bin/notary ${PREFIX}/bin/notary-signer
 	@echo "+ $@"
 
+escrow: ${PREFIX}/bin/escrow
```
Comment (endophage) [inline Makefile]: No strong feelings. It didn't seem necessary as "escrow" isn't quite as generic as "server". I mostly just got tired of typing `notary-*` on the front of everything :-P

File: trustmanager/remoteks/client_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	name := "testfile"
+	bytes := []byte{'1'}
+
+	s := setupTestServer(t)
+	defer s.Stop()
+
+	// can't just use NewRemoteStore because it correctly sets up tls
+	// config and for testing purposes it's easier for the client to just
+	// be insecure
+	cc, err := grpc.Dial(
```
Comment (ecordell) [inline trustmanager/remoteks/client_test.go]: `NewRemoteStore` could swap out `grpc.WithTransportCredentials` for `grpc.WithInsecure` if the tlsconfig is nil?

File: trustmanager/remoteks/server_test.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/docker/notary/storage"
+)
+
+func TestNewGRPCStorage(t *testing.T) {
+	s := NewGRPCStorage(nil)
+	require.IsType(t, &GRPCStorage{}, s)
+}
+
+// Set, Get, List, Remove, List, Get
+func TestGRPCStorage(t *testing.T) {
```
Comment (ecordell) [inline trustmanager/remoteks/server_test.go]: What about the error cases?

**Comment by endophage** [comment]
@ecordell the full repo will be coming in follow up PRs, with something identical or at least very, very close to that interface. This was a nice self contained piece of functionality (that is also somewhat necessary for the remote full repository).

The dependency injection PR from @n4ss is also foundational work for the full remote repo.

File: trustmanager/remoteks/client_test.go (hunk 10 above)
Comment (endophage) [inline trustmanager/remoteks/client_test.go]: You are correct but it seems like `InsecureSkipVerify` and `WithInsecure` achieve the same thing so I figured it was easier to just use the well understood tls config and let users set that up however they please, rather than having 2 ways to set clients to insecure mode.

File: Makefile (hunk 9 above)
Comment (cyli) [inline Makefile]: Should the escrow binary also be added to the `binaries` and `static` targets?

Comment (endophage) [inline Makefile]: @cyli I think once we have configurable integration with the NotaryRepository that would make sense. At the moment it doesn't really plug into anything and until @n4ss's dependency injection PR is merged, there's no way to plug it in short of forked code.

File: trustmanager/remoteks/client.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+	}, nil
+}
+
+// Set stores the data using the provided fileName
+func (s *RemoteStore) Set(fileName string, data []byte) error {
+	sm := &SetMsg{
+		FileName: fileName,
+		Data:     data,
+	}
+	_, err := s.client.Set(context.Background(), sm)
```
Comment (cyli) [inline trustmanager/remoteks/client.go]: Non-blocking - should we be setting timeouts for these by default?

File: trustmanager/remoteks/keystore.proto [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    string FileName = 1;
+}
+
+message ByteMsg {
+    bytes Data = 1;
+}
+
+message StringListMsg {
+    repeated string FileNames = 1;
+}
```
Comment (cyli) [inline trustmanager/remoteks/keystore.proto]: Non-blocking:  missing newline

File: trustmanager/remoteks/server.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+	if err != nil {
+		return &ByteMsg{}, err
+	}
+	return &ByteMsg{Data: data}, nil
+}
+
+// ListFiles returns all known identifiers in the storage backend.
+func (s *GRPCStorage) ListFiles(ctx context.Context, _ *google_protobuf.Empty) (*StringListMsg, error) {
+	lst := s.backend.ListFiles()
+	logrus.Debugf("found keys: %s", strings.Join(lst, ","))
```
Comment (cyli) [inline trustmanager/remoteks/server.go]: If the signer is eventually going to be a remote escrow, should we actually list every key in the signer in the debug logs?  Perhaps we can debug log the number?

File: trustmanager/remoteks/server.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+// backend.
+func NewGRPCStorage(backend trustmanager.Storage) *GRPCStorage {
+	return &GRPCStorage{
+		backend: backend,
+	}
+}
+
+// Set writes the provided data under the given identifier.
+func (s *GRPCStorage) Set(ctx context.Context, msg *SetMsg) (*google_protobuf.Empty, error) {
+	logrus.Debugf("storing: %s", msg.FileName)
```
Comment (cyli) [inline trustmanager/remoteks/server.go]: Non-blocking:  Should the other RPC calls also be logged?  Or at least `Remove`, since both are writes?

**Comment by cyli** [comment]
Could we also add `trustmanager/remoteks/*.pb.go` to the codecov ignore section?

File: trustmanager/remoteks/server.go (hunk 14 above)
Comment (endophage) [inline trustmanager/remoteks/server.go]: I really need an "ultra debug" mode. Listing keys has been useful but you're right that in this case it could be a lot.

File: trustmanager/remoteks/client_test.go (hunk 10 above)
Comment (ecordell) [inline trustmanager/remoteks/client_test.go]: would it be easier to just use the certs already included in the project for testing? it's not a big deal since anyone actually running this will certainly discover problems with tls config right away, but it would be nice to have it in the test path if possible.

Comment (endophage) [inline trustmanager/remoteks/client_test.go]: Pushed

**Comment by endophage** [comment]
I think I've addressed everything here. Just rebased, tests passed locally, hopefully circle agrees.

File: codecov.yml [hunk 16]
```diff
@@ -20,5 +20,6 @@ coverage:
     - "tuf/testutils/*"
     - "vendor/*"
     - "proto/*.pb.go"
+    - "trustmanager/remoteks/keystore.proto"
```
Comment (cyli) [inline codecov.yml]: Non-blocking - I think you may need to exclude the `*.pb.go` files, since the coverage will be calculated for the generated go files, not the proto files (I don't think codecov knows about protos)

**Comment by cyli** [comment]
I've re-kicked the yubikey tests, and they past.  Codecov is being weird.  :(   This otherwise LGTM, modulo a codecov ignore change (which does not account for the codecov weirdness atm)

File: codecov.yml (hunk 16 above)
Comment (endophage) [inline codecov.yml]: Oh sorry, I totally crossed my wires. I meant to exclude the pb.go!

Comment (endophage) [inline codecov.yml]: Ok fixed now. Must have been one of those occasions where I was thinking of something else while typing and the wrong thing came out.

Comment (cyli) [inline codecov.yml]: I do that at least half the time :)
